AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStore/src/KMSKeystoreOperations.dfy

```diff
@@ -630,6 +630,75 @@ module {:options "/functionSyntax:4" } KMSKeystoreOperations {
 
   }
 
+  method DecryptKeyForHv2(
+    encryptedKey: Types.EncryptedHierarchicalKey,
+    kmsConfiguration: Types.KMSConfiguration,
+    grantTokens: KMS.GrantTokenList,
+    kmsClient: KMS.IKMSClient
+  )
+    returns (output: Result<KMS.DecryptResponse, Types.Error>)
+    requires encryptedKey.EncryptionContext.Keys !! Structure.BRANCH_KEY_RESTRICTED_FIELD_NAMES
+
+    requires kmsClient.ValidState()
+    modifies kmsClient.Modifies
+    ensures kmsClient.ValidState()
+
+    ensures !KmsArn.ValidKmsArn?(encryptedKey.KmsArn) ==> output.Failure?
+    ensures !AttemptKmsOperation?(kmsConfiguration, encryptedKey.KmsArn) ==> output.Failure?
+
+    ensures output.Success?
+            ==>
+              && |kmsClient.History.Decrypt| == |old(kmsClient.History.Decrypt)| + 1
+              && AwsKmsBranchKeyDecryption?(
+                   encryptedKey,
+                   kmsConfiguration,
+                   grantTokens,
+                   kmsClient,
+                   Seq.Last(kmsClient.History.Decrypt)
+                 )
+    ensures output.Success?
+            ==>
+              && Seq.Last(kmsClient.History.Decrypt).output.Success?
+              && output.value == Seq.Last(kmsClient.History.Decrypt).output.value
+              && output.value.Plaintext.Some?
+              && 80 == |output.value.Plaintext.value|
+  {
+    :- Need(
+      && KmsArn.ValidKmsArn?(encryptedKey.KmsArn)
+         // This check is overloaded.
+         // It is incredibly unlikely that the the stored ciphertext
+         // has dropped to 0 or exceeds the KMS limit.
+         // So the error message is left unchanged.
+      && KMS.IsValid_CiphertextType(encryptedKey.CiphertextBlob),
+      Types.KeyStoreException( message := ErrorMessages.RETRIEVED_KEYSTORE_ITEM_INVALID_KMS_ARN)
+    );
+    :- Need(
+      AttemptKmsOperation?(kmsConfiguration, encryptedKey.KmsArn),
+      Types.KeyStoreException( message := ErrorMessages.GET_KEY_ARN_DISAGREEMENT)
+    );
+
+    var kmsKeyArn := GetArn(kmsConfiguration, encryptedKey.KmsArn);
+    var maybeDecryptResponse := kmsClient.Decrypt(
+      KMS.DecryptRequest(
+        CiphertextBlob := encryptedKey.CiphertextBlob,
+        EncryptionContext := Some(encryptedKey.EncryptionContext),
```

Comment:
Blocking: This needs to be the Extracted EC

Reply:
Not really. DecryptRequest needs Option EC
```
datatype DecryptRequest = | DecryptRequest (
    nameonly CiphertextBlob: CiphertextType ,
    nameonly EncryptionContext: Option<EncryptionContextType> := Option.None ,
    nameonly GrantTokens: Option<GrantTokenList> := Option.None ,
    nameonly KeyId: Option<KeyIdType> := Option.None ,
    nameonly EncryptionAlgorithm: Option<EncryptionAlgorithmSpec> := Option.None ,
    nameonly Recipient: Option<RecipientInfo> := Option.None ,
    nameonly DryRun: Option<NullableBooleanType> := Option.None
  )
```

Reply:
Err... yes, the KMS Request is valid.
But it is not correct for HV-2.
```
Types.EncryptedHierarchicalKey
```
is defined by the Storage layer;
so `encryptedKey.EncryptionContext` is going to be the Branch Key Context,
as compared to the Branch Key's Encryption Context;
unless we refactor Storage
which maybe an option.

Reply:
Still WIP but I decided to send ciphertextBlob, encryptionContextToKms, and kmsArnFromStorage instead of Types.EncryptedHierarchicalKey
    
https://github.com/aws/aws-cryptographic-material-providers-library/pull/1368/files#diff-0e12585873434d898794557b0e9710e9e827e6569ad433fcd8dd7750b32701e2R636-R638

Reply:
>Still WIP but I decided to send ciphertextBlob, encryptionContextToKms, and kmsArnFromStorage instead of Types.EncryptedHierarchicalKey

Because it turned out to be very confusing with the membershape named as EC